Extracts data from snomed

sigal23 opened the pull request.

sigal23 committed:
dump dicts to files

sigal23 committed:
Keeping concepts without parent

aryehgigi commented on main.py:
```diff
@@ -1,3 +1,4 @@
+import json
 
 if __name__ == '__main__':
 
```
Consider moving aliasesFields to be a constant out of the scope of the function. not critical but usually hardcoded values are stored in one place in the beginning of the file or in a different file named common or constants etc.

aryehgigi replied:
also move the id's themselves to constants, than you can use them all over the file. maybe also the '1' on line 19, you can change to ACTIVE = '1'

aryehgigi commented on main.py:
```diff
@@ -1,3 +1,4 @@
+import json
 
 if __name__ == '__main__':
 
```
consider using the csv package
you can than do:
```
row_iter = csv.DictReader(f)
for row in row_iter:
    # use named keys to access row
    # e.g. row["id"]
```
this would be much more readable.
notice that you might need to pass a "delimiter" option when initializing DictReader

aryehgigi commented on main.py:
```diff
@@ -32,19 +33,34 @@
                 elif row[6] == aliasesFields['fully specified name']:
```
instead of 4 if's you can have nested if's. you can share the conditions where appropriate.

aryehgigi commented on main.py:
```diff
@@ -32,19 +33,34 @@
+        concept_without_father = list(concept_dict.keys())
         for line in relation:
             row = line.strip().split('\t')
             if row[2] == '1' and row[7] == '116680003':
```
Store as constants

aryehgigi commented on main.py:
```diff
@@ -32,19 +33,34 @@
+                if row[4] in concept_without_father:
+                    concept_without_father.remove(row[4])
+
+    # for k in concept_dict.keys():
```
remove redundant code

aryehgigi commented on main.py:
```diff
@@ -32,19 +33,34 @@
 
-    for k in is_a_dict.keys():
-        print(str(k) + ' ' + str(is_a_dict[k]))
+    # loaded_json = json.load(f)
```
remove redundant code

aryehgigi commented on main.py:
```diff
@@ -1,3 +1,4 @@
+import json
```
General comments:
1. you should split the code to main, extract and convert-to-our json. notice that you use data files and create json intermediate files that belong to a specific format, they shouldn't be sitting in the main directory. instead create a folder for the extract code and its data.
2. document the code - so future Sigal won't be mad at you

sigal23 committed:
add gitignore

Reference to the original files
Correction of conditions and reference only to the concepts in concept_dict
add tqdm

sigal23 committed:
Correction of code according to the review

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+                if row['typeId'] == SYNONYM:
+                    if row['id'] in prefer_terms:
+                        concept_dict[row['conceptId']]['preferred synonym'].append(row['term'])
+                        # The name of each concept will be the preferred synonym
```
fix the documentation here: `The canonical name...`

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+                    concept_without_father.remove(row['sourceId'])
+
+    # Saving the relevant data we extracted to json files
+    with open("snomed/concepts_test.json", 'w') as f:
```
i didnt notice before but i think we can remove the `test` from the name.. (and fix the names of the files we already extracted to be accordingly)

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+    with open("snomed/concepts_test.json", 'w') as f:
+        json.dump(concept_dict, f)
+
+    with open("snomed/is_a_test.json", 'w') as f:
```
see last comment

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+    with open("snomed/is_a_test.json", 'w') as f:
+        json.dump(is_a_dict, f)
+
+    with open("snomed/concept_without_father_test.json", 'w') as f:
```
see last comment

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+        json.dump(is_a_dict, f)
+
+    with open("snomed/concept_without_father_test.json", 'w') as f:
+        json.dump(concept_without_father, f)
```
i think you can already add a return statement with the values you want to pass out (after this line of course)

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def snomed_extract():
+
```
Very minor, but this space is redundant :)

aryehgigi commented on snomed/snomed_extract.py:
```diff
@@ -0,0 +1,82 @@
+    with open(DESCRIPTION) as desc:
+        row_iter = csv.DictReader(desc, delimiter='\t')
+        for row in tqdm(row_iter):
+            if row['active'] == ACTIVE and row['languageCode'] == 'en' and row['conceptId'] in concept_dict:
```
Not critical, but i think the `'en'` could also be in the constants.. as it is another content that they might change in the future

sigal23 committed:
Create a json file in our format

Correction of code according to the review

aryehgigi commented on snomed/README.md:
```diff
@@ -0,0 +1,23 @@
+
+## Download relevant files
+
+Navigate to https://www.nlm.nih.gov/healthit/snomedct/us_edition.html and download the file.
```
1. Maybe add a comment on the US edition `We chose to use this edition as it extends over the official international edition  and should be most updated and full`.
2. its not exactly clear which file to do download, maybe a more concrete link to the download, or a name of the file to download?

aryehgigi commented on snomed/README.md:
```diff
@@ -0,0 +1,23 @@
+1.  The file: "sct2_Concept_Snapshot_US1000124_20220301.txt" from SnomedCT_USEditionRF2_PRODUCTION_20220301T120000Z/Snapshot/Terminology/
+2.  The file: "sct2_Description_Snapshot-en_US1000124_20220301.txt" from SnomedCT_USEditionRF2_PRODUCTION_20220301T120000Z/Snapshot/Terminology/
+3.  The file: "sct2_Relationship_Snapshot_US1000124_20220301.txt" from SnomedCT_USEditionRF2_PRODUCTION_20220301T120000Z/Snapshot/Terminology/
+4.  The file: "der2_cRefset_LanguageSnapshot-en_US1000124_20220301.txt" from /home/sigal/Downloads/SnomedCT_USEditionRF2_PRODUCTION_20220301T120000Z/Snapshot/Refset/Language/
```
I guess youve added this by mistake :) (i refer to `/home/sigal/Downloads/`)

aryehgigi commented on convert_to_our_format.py:
```diff
@@ -0,0 +1,50 @@
+import json
+
+aliasesFields = ['preferred synonym', 'synonyms', 'preferred fully specified name', 'fully specified names']
+f = open('snomed/concepts.json')
```
lines 4-9 should be in the `create_our_format` function, as you dont want this to happen in anyway (what happens of there is no such file?). but only inside an IF condition.
if you wanted it to be global so the recursive function could use it, you can:
1. pass the argument to the recursive call, OR,
2. define the recursive function itself inside the `create_our_format` function so it will be aware of these params

aryehgigi commented on convert_to_our_format.py:
```diff
@@ -0,0 +1,50 @@
+        concept_leaf = {"name": concept_dict[conc_id]["name"], "id": id, "state": {"aliases": aliases}, "children": []}
+        return concept_leaf
+
+    aliases = {}
```
this code (lines 24-28) is exactly the same as before (lines 17-21)
you could extract the code to a function and reuse it. but there is a more neat solution:
if you had this code only once, it can still be used for both "the stopping condition" and "middle step in the recursion". Just pass the children to the children field, and automatically the recursion will not be happening for leavse because they dont have children. Imagine you remove lines 16-22 and 30-32 and replace the ending of line 28 with:
`"children":  [create_hierarchy(c) for c in is_a_dict.get(conc_id, [])]` notice that in the case there are no children (a leaf) the `get` returns an empty list, so the list comprehension wont run and no call to create_hierarchy will happen => the "children" field would be an empty list, exactly as you want.

aryehgigi commented on convert_to_our_format.py:
```diff
@@ -0,0 +1,50 @@
+    return concept
+
+
+def create_our_format():
```
its a bit confusing that you call this file convert and the function create. which is it?
you can choose either, but (a) be consistent, (b) if you go with `convert` it means you need to return a value from this function (convert implies you get x and return a converted y) and not create here anything (leave that to the main file/function) but if you choose to go with `create` then its ok as you did.

aryehgigi commented on convert_to_our_format.py:
```diff
@@ -0,0 +1,50 @@
+
+    root = {"name": "root", "id": 1, "state": {"aliases": {aliasesFields[0]: ["all"]}}, "children": []}
+
+    if len(concept_without_father) == 1:
```
remember we said this code needs to be generic. maybe in other taxonomies we would want to keep their top single root to be under our root?
instead we need to behave here equally to any format = simply do what you did in the `else` block of code - and remove the entire `if` part. then in the snomed_extract code you should this logic where you remove the head and return its immediate children as the list of heads
the place you know you want to do it - thats where it should happen :)

aryehgigi commented on convert_to_our_format.py:
```diff
@@ -0,0 +1,50 @@
+        root["children"] = [create_hierarchy(c) for c in concept_without_father]
+
+    with open("snomed/snomed_our_format.json", "w") as fi:
+        json.dump({"nodes": [root], "aliasesFields": aliasesFields}, fi)
```
general comment for this file: it looks good overall! fix the few comments i left above, and dont foget to document the code a bit :)

sigal23 committed:
Correction of code according to the review

sigal23 committed:
Turning the 'is_a' graph into a tree

sigal23 committed:
Code documentation + README

aryehgigi commented on main.py:
```diff
@@ -1,8 +1,38 @@
+    parser = argparse.ArgumentParser()
+    # Add an arguments
+    parser.add_argument('--extract', default=False, action='store_true')
+    parser.add_argument('--no-extract', dest='extract', action='store_false')
```
isnt this no-extract option redundant?

aryehgigi reviewed (approved):
Looks very good :)
fix the single comment that i left, and you can merge this PR

sigal23 committed:
fix args
